### ISSUE TITLE
Add `batch.large_event` option

### ISSUE THREAD

**Issue by ktff** [timberio/vector#8705]
### Use-cases

Currently when batching encoded events that are larger than configured `batch.max_bytes` they will be dropped. Which is fine when there is a max size of request that the downstream service is willing to accept, but `batch.large_event` option would enable users to not loose events when that isn't the case.  

### Proposal
Add enum option `batch.large_event` with two values:
* `drop`, default 
* `send`, with some documentation warning

This option would be added for all sinks that have batching. 

The default value would be `drop`, and while there may be some services for which `send` would make more sense, `drop` would prevent sending of some absurd events by default.

### References
https://github.com/timberio/vector/pull/2866#issuecomment-646891090
https://github.com/timberio/vector/pull/2866#issuecomment-646892252
#7395

**Comment by lukesteensen** [comment]
No, this isn't part of failure branching, at least as it exists now. We only know when an event is too large when it's already in a sink being encoded. We could maybe add a feature to dead-letter these from within the sink itself, which would reuse some of the same internals but still require a reasonable amount of design on its own. Sinks having any outputs at all is a large-ish conceptual change.

**Comment by pgassmann** [comment]
@spencergilbert @jdrouet Thank you. the loki sink rewrite in the nightly version 0.18.0 solved the issue. Before, even with debug and trace logging, I found no error/explanation for the sudden stop of logs sent to loki.

**Comment by jszwedko** [comment]
@pgassmann Good to hear!

The issue you observed was essentially that the state machine controlling processing in the `loki` sink would get into a state where it would just stop polling (which is why you never saw any logs around the halt). We spent a significant amount of time debugging an issue in a similarly architected sink, `aws_s3`, and came to the conclusion that the current sink architecture was too complex to narrow down where the issue was and so opted to rewrite the sink in a more straight-forward fashion. We are migrating all of the other sinks to this new pattern which included the `loki` sink recently.
